mesh/mesh-adapter-tcpserver.c

```diff
@@ -0,0 +1,516 @@
+
+	return true;
+}
+
+static char *create_dbus_path(uint16_t port)
+{
+	return l_strdup_printf("%s/tcp_%d", BLUEZ_MESH_PATH, port);
+}
+
+void *mesh_adapter_tcpserver_new(uint16_t port)
+{
+	struct mesh_adapter_tcpserver * adapter;
+
+	adapter = l_new(struct mesh_adapter_tcpserver, 1);
+
+	adapter->adapter.api.dbus_init = mesh_adapter_tcpserver_dbus_init;
+	adapter->dbus_path = create_dbus_path(port);
+	adapter->acl = l_queue_new();
+
+	mesh_adapter_tcpserver_conf_load(conf_path, adapter->acl);
+
+	return adapter;
+}
+
+uint8_t *mesh_adapter_tcpserver_get_psk(const char *identity, void *user_data)
+{
+	struct mesh_adapter_tcpserver * adapter = user_data;
+	struct acl_entry *entry;
+
+	unsigned char *id = l_util_from_hexstring(identity, NULL);
+
+	entry = acl_entry_find_by_identity(adapter->acl, id);
+
+	l_free(id);
+
+	return entry ? entry->psk : NULL;
+}
+
+
+// Configuration storage part
+static bool load_acl(json_object *jobj, struct l_queue *acl)
+{
+	json_object *jarray;
+	json_object *jentry;
+	json_object *jval;
+	int key_cnt;
+
+	json_object_object_get_ex(jobj, "ACL", &jarray);
+	if (!jarray || json_object_get_type(jarray) != json_type_array)
+		return false;
+
+	l_info("mam acl :)");
+
+	key_cnt = json_object_array_length(jarray);
+	if (key_cnt < 0)
+		return false;
+
+	for (int i = 0; i < key_cnt; ++i) {
+		struct acl_entry *entry = l_new(struct acl_entry, 1);
+		const char *str;
+		uint8_t *val;
+
+		jentry = json_object_array_get_idx(jarray, i);
+
+		if (!json_object_object_get_ex(jentry, "token", &jval))
+			return false;
+
+		str = json_object_get_string(jval);
+
+		val = l_util_from_hexstring(str, NULL);
+
+		entry->token = *(uint64_t *)val;
+
+		l_info("token: %ju", entry->token);
+
+		if (!json_object_object_get_ex(jentry, "identity", &jval))
+			return false;
+
+		str = json_object_get_string(jval);
+
+		val = l_util_from_hexstring(str, NULL);
+		memcpy(entry->identity, val, IDENTITY_LEN);
+		l_free(val);
+
+		l_info("identity: '%s'", str);
+
+		if (!json_object_object_get_ex(jentry, "psk", &jval))
+			return false;
+
+		str = json_object_get_string(jval);
+
+		val = l_util_from_hexstring(str, NULL);
+		memcpy(entry->psk, val, PSK_LEN);
+		l_free(val);
+
+		l_info("psk: '%s'", str);
+
+		l_queue_push_tail(acl, entry);
+	}
+
+	return true;
+}
+
+
+bool mesh_adapter_tcpserver_conf_load(const char *fname, struct l_queue *acl)
+{
+	int fd;
+	char *str;
+	struct stat st;
+	ssize_t sz;
+	json_object *jcfg;
+
+	fd = open(fname, O_RDONLY);
+	if (fd < 0)
+		return false;
+
+	if (fstat(fd, &st) == -1) {
+		close(fd);
+		return false;
+	}
+
+	str = l_new(char, st.st_size + 1);
+
+	sz = read(fd, str, st.st_size);
+	if (sz != st.st_size) {
+		l_error("Failed to read configuration file %s", fname);
+		return false;
+	}
+
+	jcfg = json_tokener_parse(str);
+
+	close(fd);
+	l_free(str);
+
+	if (!jcfg)
+		return false;
+
+	if (!load_acl(jcfg, acl))
+		return false;
+
+	return true;
+}
+
+static bool add_token(json_object *jobj, uint64_t token)
```

Comment:
I think you can keep the token simply as a uint64 number, without hex encoding.

Reply:
`uint64_t` is not supported by JSON-c library. It is possible to store token as `int64_t` or hexstr. Should I change it to store it as `int64_t`?

Reply:
Alright, let's leave it as hexstring then.